lib/engine-universal/src/artifact.rs

```diff
@@ -1,402 +1,211 @@
 //! Define `UniversalArtifact` to allow compiling and instantiating to be
 //! done as separate steps.
 
-use crate::engine::{UniversalEngine, UniversalEngineInner};
-use crate::link::link_module;
-#[cfg(feature = "compiler")]
-use crate::serialize::SerializableCompilation;
-use crate::serialize::SerializableModule;
-use enumset::EnumSet;
 use loupe::MemoryUsage;
+use std::collections::BTreeMap;
+use std::convert::TryFrom;
 use std::sync::{Arc, Mutex};
-use wasmer_compiler::{CompileError, CpuFeature, Features, Triple};
-#[cfg(feature = "compiler")]
-use wasmer_compiler::{CompileModuleInfo, ModuleEnvironment, ModuleMiddlewareChain};
-use wasmer_engine::{
-    register_frame_info, Artifact, DeserializeError, FunctionExtent, GlobalFrameInfoRegistration,
-    SerializeError,
-};
-#[cfg(feature = "compiler")]
-use wasmer_engine::{Engine, Tunables};
-use wasmer_types::entity::{BoxedSlice, PrimaryMap};
+use wasmer_compiler::Triple;
+use wasmer_engine::{GlobalFrameInfoRegistration, InstantiationError, RuntimeError};
+use wasmer_types::entity::{BoxedSlice, EntityRef, PrimaryMap};
 use wasmer_types::{
-    FunctionIndex, LocalFunctionIndex, MemoryIndex, ModuleInfo, OwnedDataInitializer,
-    SignatureIndex, TableIndex,
+    DataIndex, ElemIndex, EntityCounts, FunctionIndex, GlobalInit, GlobalType, LocalFunctionIndex,
+    LocalGlobalIndex, MemoryType, OwnedDataInitializer, OwnedTableInitializer, SignatureIndex,
+    TableType,
 };
 use wasmer_vm::{
-    FuncDataRegistry, FunctionBodyPtr, MemoryStyle, TableStyle, VMSharedSignatureIndex,
-    VMTrampoline,
+    Artifact, FunctionBodyPtr, FunctionExtent, InstanceHandle, MemoryStyle, Resolver, TableStyle,
+    Tunables, VMImport, VMImportType, VMLocalFunction, VMOffsets, VMSharedSignatureIndex,
 };
 
-const SERIALIZED_METADATA_LENGTH_OFFSET: usize = 22;
-const SERIALIZED_METADATA_CONTENT_OFFSET: usize = 32;
-
-/// A compiled wasm module, ready to be instantiated.
+/// A compiled wasm module, containing everything necessary for instantiation.
 #[derive(MemoryUsage)]
 pub struct UniversalArtifact {
-    serializable: SerializableModule,
-    finished_functions: BoxedSlice<LocalFunctionIndex, FunctionBodyPtr>,
-    #[loupe(skip)]
-    finished_function_call_trampolines: BoxedSlice<SignatureIndex, VMTrampoline>,
-    finished_dynamic_function_trampolines: BoxedSlice<FunctionIndex, FunctionBodyPtr>,
-    signatures: BoxedSlice<SignatureIndex, VMSharedSignatureIndex>,
-    func_data_registry: Arc<FuncDataRegistry>,
-    frame_info_registration: Mutex<Option<GlobalFrameInfoRegistration>>,
-    finished_function_lengths: BoxedSlice<LocalFunctionIndex, usize>,
+    // TODO: figure out how to allocate fewer distinct structures onto heap. Maybe have an arena…?
+    pub(crate) engine: crate::UniversalEngine,
+    pub(crate) import_counts: EntityCounts,
+    pub(crate) start_function: Option<FunctionIndex>,
+    pub(crate) vmoffsets: VMOffsets,
+
+    #[loupe(skip)] // TODO(0-copy): support loupe...
```

Comment:
Should either remove loupe entirely or fix all the places where this is being skipped.